radixdlt-java/src/main/java/com/radixdlt/client/core/atoms/particles/Particle.java

```diff
@@ -1,9 +1,88 @@
 package com.radixdlt.client.core.atoms.particles;
 
+import com.radixdlt.client.core.atoms.particles.quarks.Quark;
 import com.radixdlt.client.core.crypto.ECPublicKey;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
 import java.util.Set;
+import java.util.stream.Stream;
+
+public abstract class Particle {
+	private final Spin spin;
+	private final List<Quark> quarks;
+
+	protected Particle(Spin spin, Quark... quarks) {
+		this.spin = spin;
+		this.quarks = new ArrayList<>(Arrays.asList(quarks));
+	}
+
+	protected Particle(Spin spin, List<Quark> quarks) {
+		this.spin = spin;
+		this.quarks = quarks;
+	}
+
+	public final Spin getSpin() {
+		return this.spin;
+	}
+
+	public abstract Set<ECPublicKey> getAddresses();
+
+	/// Methods taken from Particle.java in Core
+
+	public final <T extends Quark> T getQuark(Class<T> type) {
+		if (type == null) {
+			throw new IllegalArgumentException("Type is null");
+		}
+
+		if (this.quarks != null) {
+			for (Quark quark : this.quarks) {
+				if (type.isAssignableFrom(quark.getClass())) {
+					return (T) quark;
+				}
+			}
+		}
+
+		return null;
+	}
+
+	public final List<Quark> getQuarks() {
+		return Collections.unmodifiableList(this.quarks);
+	}
+
+	public final <T extends Quark> List<T> getQuarks(Class<T> type) {
+		if (type == null) {
+			throw new IllegalArgumentException("Type is null");
+		}
+
+		List<Quark> quarks = new ArrayList<>();
+
+		if (this.quarks != null) {
+			for (Quark quark : this.quarks) {
+				if (type.isAssignableFrom(quark.getClass())) {
```

Comment:
I think there should be a constraint that all quarks are final classes in which case we should just use a strict equals in this comparison